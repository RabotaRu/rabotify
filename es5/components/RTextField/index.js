'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _RTextField = require('./RTextField.vue');

var _RTextField2 = _interopRequireDefault(_RTextField);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/* istanbul ignore next */
_RTextField2.default.install = function install(Vue) {
  Vue.component('r-text-field', _RTextField2.default);
};

exports.default = _RTextField2.default;