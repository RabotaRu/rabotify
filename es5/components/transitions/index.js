'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.RExpandTransition = exports.RScaleTransitionGroup = exports.RSlideYReverseTransitionGroup = exports.RSlideYTransitionGroup = exports.RSlideXReverseTransitionGroup = exports.RSlideXTransitionGroup = exports.RFadeTransitionGroup = exports.RSlideYReverseTransition = exports.RSlideYTransition = exports.RSlideXReverseTransition = exports.RSlideXTransition = exports.RScaleTransition = exports.RFadeTransition = exports.RToastTransition = exports.RFabTransition = exports.RMenuTransition = exports.RTabReverseTransition = exports.RTabTransition = exports.RCarouselReverseTransition = exports.RCarouselTransition = exports.RBottomSheetTranstion = exports.RDialogRightTransition = exports.RDialogLeftTransition = exports.RDialogBottomTransition = exports.RDialogTopTransition = exports.RDialogTransition = undefined;

var _helpers = require('../../util/helpers');

var _expandTransition = require('./expand-transition');

var _expandTransition2 = _interopRequireDefault(_expandTransition);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Component specific transitions
var RDialogTransition = exports.RDialogTransition = (0, _helpers.createSimpleTransition)('dialog-transition');
var RDialogTopTransition = exports.RDialogTopTransition = (0, _helpers.createSimpleTransition)('dialog-top-transition');
var RDialogBottomTransition = exports.RDialogBottomTransition = (0, _helpers.createSimpleTransition)('dialog-bottom-transition');
var RDialogLeftTransition = exports.RDialogLeftTransition = (0, _helpers.createSimpleTransition)('dialog-left-transition');
var RDialogRightTransition = exports.RDialogRightTransition = (0, _helpers.createSimpleTransition)('dialog-right-transition');

var RBottomSheetTranstion = exports.RBottomSheetTranstion = (0, _helpers.createSimpleTransition)('bottom-sheet-transition');
var RCarouselTransition = exports.RCarouselTransition = (0, _helpers.createSimpleTransition)('carousel-transition');
var RCarouselReverseTransition = exports.RCarouselReverseTransition = (0, _helpers.createSimpleTransition)('carousel-reverse-transition');
var RTabTransition = exports.RTabTransition = (0, _helpers.createSimpleTransition)('tab-transition');
var RTabReverseTransition = exports.RTabReverseTransition = (0, _helpers.createSimpleTransition)('tab-reverse-transition');
var RMenuTransition = exports.RMenuTransition = (0, _helpers.createSimpleTransition)('menu-transition');
var RFabTransition = exports.RFabTransition = (0, _helpers.createSimpleTransition)('fab-transition', 'center center', 'out-in');
var RToastTransition = exports.RToastTransition = (0, _helpers.createSimpleTransition)('toast-transition', 'center center', 'out-in');

// Generic transitions
var RFadeTransition = exports.RFadeTransition = (0, _helpers.createSimpleTransition)('fade-transition');
var RScaleTransition = exports.RScaleTransition = (0, _helpers.createSimpleTransition)('scale-transition');
var RSlideXTransition = exports.RSlideXTransition = (0, _helpers.createSimpleTransition)('slide-x-transition');
var RSlideXReverseTransition = exports.RSlideXReverseTransition = (0, _helpers.createSimpleTransition)('slide-x-reverse-transition');
var RSlideYTransition = exports.RSlideYTransition = (0, _helpers.createSimpleTransition)('slide-y-transition');
var RSlideYReverseTransition = exports.RSlideYReverseTransition = (0, _helpers.createSimpleTransition)('slide-y-reverse-transition');

// Generic transitions' groups
var RFadeTransitionGroup = exports.RFadeTransitionGroup = (0, _helpers.createSimpleTransitionGroup)('fade-transition');
var RSlideXTransitionGroup = exports.RSlideXTransitionGroup = (0, _helpers.createSimpleTransitionGroup)('slide-x-transition');
var RSlideXReverseTransitionGroup = exports.RSlideXReverseTransitionGroup = (0, _helpers.createSimpleTransitionGroup)('slide-x-reverse-transition');
var RSlideYTransitionGroup = exports.RSlideYTransitionGroup = (0, _helpers.createSimpleTransitionGroup)('slide-y-transition');
var RSlideYReverseTransitionGroup = exports.RSlideYReverseTransitionGroup = (0, _helpers.createSimpleTransitionGroup)('slide-y-reverse-transition');
var RScaleTransitionGroup = exports.RScaleTransitionGroup = (0, _helpers.createSimpleTransitionGroup)('scale-transition');

// JavaScript transitions
var RExpandTransition = exports.RExpandTransition = (0, _helpers.createJavaScriptTransition)('expand-transition', (0, _expandTransition2.default)());

/* istanbul ignore next */
// eslint-disable-next-line max-statements
function install(Vue) {
  // Component transitions
  Vue.component('r-bottom-sheet-transition', RBottomSheetTranstion);
  Vue.component('r-carousel-transition', RCarouselTransition);
  Vue.component('r-carousel-reverse-transition', RCarouselReverseTransition);
  Vue.component('r-dialog-transition', RDialogTransition);
  Vue.component('r-dialog-top-transition', RDialogTopTransition);
  Vue.component('r-dialog-bottom-transition', RDialogBottomTransition);
  Vue.component('r-dialog-left-transition', RDialogLeftTransition);
  Vue.component('r-dialog-right-transition', RDialogRightTransition);
  Vue.component('r-tab-reverse-transition', RTabReverseTransition);
  Vue.component('r-tab-transition', RTabTransition);
  Vue.component('r-fab-transition', RFabTransition);
  Vue.component('r-menu-transition', RMenuTransition);

  // Generic transitions
  Vue.component('r-fade-transition', RFadeTransition);
  Vue.component('r-scale-transition', RScaleTransition);
  Vue.component('r-slide-x-transition', RSlideXTransition);
  Vue.component('r-slide-x-reverse-transition', RSlideXReverseTransition);
  Vue.component('r-slide-y-transition', RSlideYTransition);
  Vue.component('r-slide-y-reverse-transition', RSlideYReverseTransition);

  // Generic group transitions
  Vue.component('r-fade-transition-group', RFadeTransitionGroup);
  Vue.component('r-slide-x-transition-group', RSlideXTransitionGroup);
  Vue.component('r-slide-x-reverse-transition-group', RSlideXReverseTransitionGroup);
  Vue.component('r-slide-y-transition-group', RSlideYTransitionGroup);
  Vue.component('r-slide-y-reverse-transition-group', RSlideYReverseTransitionGroup);
  Vue.component('r-scale-transition-group', RScaleTransitionGroup);

  // JavaScript transitions
  Vue.component('r-expand-transition', RExpandTransition);
}

exports.default = install;