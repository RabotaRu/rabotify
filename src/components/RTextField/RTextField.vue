<script>
  import Colorable from '../../mixins/colorable';
  import Input from '../../mixins/input';
  import CleaveMaskable from '../../mixins/cleave-maskable';

  export default {
    name: 'r-text-field',

    mixins: [
      Colorable,
      Input,
      CleaveMaskable
    ],

    inheritAttrs: false,

    data () {
      return {
        initialValue: null,
        inputHeight: null,
        internalChange: false,
        badInput: false,
        lazySelection: 0
      };
    },

    props: {
      autofocus: Boolean,
      autoGrow: Boolean,
      box: Boolean,
      clearable: Boolean,
      color: String,
      counter: [
        Boolean,
        Number,
        String
      ],
      fullWidth: Boolean,
      multiLine: Boolean,
      placeholder: String,
      prefix: String,
      rows: {
        default: 5
      },
      singleLine: Boolean,
      solo: Boolean,
      suffix: String,
      textarea: Boolean,
      noResize: Boolean,
      small: Boolean,
      large: Boolean,
      type: {
        type: String,
        default: 'text'
      }
    },

    computed: {
      classes () {
        const classes = {
          'input-group_text-field': true,
          'input-group_text-field-box': this.box,
          'input-group_single-line': this.singleLine || this.solo,
          'input-group_solo': this.solo,
          'input-group_multi-line': this.multiLine,
          'input-group_full-width': this.fullWidth,
          'input-group_prefix': this.prefix,
          'input-group_suffix': this.suffix,
          'input-group_textarea': this.textarea,
          'input-group_no-resize': this.noResize || this.autoGrow
        };

        if (this.hasError) {
          classes['text_error'] = true;
        } else {
          return this.addTextColorClassChecks(classes);
        }

        return classes;
      },
      inputLength () {
        if (this.inputValue) {
          return this.inputValue.toString().length;
        }
        return 0;
      },
      count () {
        return `${this.inputLength} / ${this.counterLength}`;
      },
      counterLength () {
        const parsedLength = parseInt(this.counter, 10);
        return isNaN(parsedLength) ? 25 : parsedLength;
      },
      inputValue: {
        get () {
          return this.lazyValue;
        },
        set (val) {
          if (this.cleaveMask && typeof window !== 'undefined') {
            this.$nextTick(_ => {
              this.lazyValue = val ? this.cleave.getFormattedValue() : val;
              this.$emit('input', this.lazyValue);
            });
          } else {
            this.lazyValue = val;
            this.$emit('input', this.lazyValue);
          }
        }
      },
      isDirty () {
        return this.lazyValue != null &&
          this.lazyValue.toString().length > 0 ||
          this.badInput ||
          [ 'time', 'date', 'datetime-local', 'week', 'month' ].includes(this.type);
      },
      shouldAutoGrow () {
        return (this.multiLine || this.textarea) && this.autoGrow;
      }
    },

    watch: {
      isFocused (val) {
        if (val) {
          this.initialValue = this.lazyValue;
        } else if (this.initialValue !== this.lazyValue) {
          this.$emit('change', this.lazyValue);
        }
      },
      value (val) {
        if (this.cleaveMask && typeof window !== 'undefined') {
          this.cleave.setRawValue(val);
          this.lazyValue = this.cleave.getFormattedValue();
        } else {
          this.lazyValue = val;
        }

        if (this.internalChange) {
          this.internalChange = false;
        }

        !this.validateOnBlur && this.validate();
        this.shouldAutoGrow && this.calculateInputHeight();
      }
    },

    mounted () {
      this.shouldAutoGrow && this.calculateInputHeight();
      this.autofocus && this.focus();
    },

    methods: {
      calculateInputHeight () {
        this.inputHeight = null;

        this.$nextTick(() => {
          const height = this.$refs.input
            ? this.$refs.input.scrollHeight
            : 0;
          const minHeight = this.rows * 24;
          const inputHeight = height < minHeight ? minHeight : height;
          this.inputHeight = inputHeight + (this.textarea ? 4 : 0);
        });
      },
      onInput (e) {
        this.inputValue = e.target.value;
        this.badInput = e.target.validity && e.target.validity.badInput;
        this.shouldAutoGrow && this.calculateInputHeight();
      },
      blur (e) {
        this.isFocused = false;
        // Reset internalChange state
        // to allow external change
        // to persist
        this.internalChange = false;

        this.$nextTick(() => {
          this.validate();
        });
        this.$emit('blur', e);
      },
      focus (e) {
        if (!this.$refs.input) {
          return;
        }

        this.isFocused = true;
        if (document.activeElement !== this.$refs.input) {
          this.$refs.input.focus();
        }
        this.$emit('focus', e);
      },
      keyDown (e) {
        // Prevents closing of a
        // dialog when pressing
        // enter or esc
        if (this.isFocused) {
          const [ escKey, enterKey ] = [ 27, 13 ];
          if (e.keyCode === escKey ||
            this.multiLine && e.keyCode === enterKey
          ) {
            e.stopPropagation();
          }
        }

        this.$emit('keydown', e);

        this.internalChange = true;
      },
      genCounter () {
        const errorCounter = this.hasError || this.inputLength >= this.counterLength;
        const warnCounter = !errorCounter && (
          Math.abs(this.inputLength - this.counterLength) / this.counterLength < 0.2
        );
        return this.$createElement('div', {
          'class': {
            'input-group__counter': true,
            'input-group__counter_error': errorCounter,
            'input-group__counter_warn': warnCounter
          }
        }, this.count);
      },
      genInput () {
        const tag = this.multiLine || this.textarea ? 'textarea' : 'input';
        const listeners = Object.assign({}, this.$listeners);
        delete listeners['change']; // Change should not be bound externally

        const data = {
          style: {},
          domProps: {
            autofocus: this.autofocus,
            disabled: this.disabled,
            required: this.required,
            value: this.lazyValue
          },
          attrs: {
            ...this.$attrs,
            readonly: this.readonly,
            tabindex: this.tabindex,
            'aria-label': (!this.$attrs || !this.$attrs.id) && this.label // Label `for` will be set if we have an id
          },
          on: Object.assign(listeners, {
            blur: this.blur,
            input: this.onInput,
            focus: this.focus,
            keydown: this.keyDown
          }),
          ref: 'input'
        };

        if (this.shouldAutoGrow) {
          data.style.height = this.inputHeight && `${this.inputHeight}px`;
        }

        if (this.placeholder) {
          data.domProps.placeholder = this.placeholder;
        }

        if (!this.textarea && !this.multiLine) {
          data.domProps.type = this.type;
        } else {
          data.domProps.rows = this.rows;
        }

        const children = [ this.$createElement(tag, data) ];

        this.prefix && children.unshift(this.genFix('prefix'));
        this.suffix && children.push(this.genFix('suffix'));

        return children;
      },
      genFix (type) {
        return this.$createElement('span', {
          'class': `input-group_text-field__${type}`
        }, this[type]);
      },
      clearableCallback () {
        this.inputValue = null;
        this.$nextTick(() => this.$refs.input.focus());
      }
    },

    render () {
      return this.genInputGroup(this.genInput(), {
        attrs: {
          tabindex: false
        }
      });
    }
  };
</script>

<style lang="scss">
  @import "../../styles/components/_text-field.scss";
</style>
