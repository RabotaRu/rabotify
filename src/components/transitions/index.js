import {
  createSimpleTransition,
  createSimpleTransitionGroup,
  createJavaScriptTransition
} from '../../util/helpers';

import ExpandTransitionGenerator from './expand-transition';

// Component specific transitions
export const RDialogTransition = createSimpleTransition('dialog-transition');
export const RDialogTopTransition = createSimpleTransition('dialog-top-transition');
export const RDialogBottomTransition = createSimpleTransition('dialog-bottom-transition');
export const RDialogLeftTransition = createSimpleTransition('dialog-left-transition');
export const RDialogRightTransition = createSimpleTransition('dialog-right-transition');

export const RBottomSheetTranstion = createSimpleTransition('bottom-sheet-transition');
export const RCarouselTransition = createSimpleTransition('carousel-transition');
export const RCarouselReverseTransition = createSimpleTransition('carousel-reverse-transition');
export const RTabTransition = createSimpleTransition('tab-transition');
export const RTabReverseTransition = createSimpleTransition('tab-reverse-transition');
export const RMenuTransition = createSimpleTransition('menu-transition');
export const RFabTransition = createSimpleTransition('fab-transition', 'center center', 'out-in');
export const RToastTransition = createSimpleTransition('toast-transition', 'center center', 'out-in');

// Generic transitions
export const RFadeTransition = createSimpleTransition('fade-transition');
export const RScaleTransition = createSimpleTransition('scale-transition');
export const RSlideXTransition = createSimpleTransition('slide-x-transition');
export const RSlideXReverseTransition = createSimpleTransition('slide-x-reverse-transition');
export const RSlideYTransition = createSimpleTransition('slide-y-transition');
export const RSlideYReverseTransition = createSimpleTransition('slide-y-reverse-transition');

// Generic transitions' groups
export const RFadeTransitionGroup = createSimpleTransitionGroup('fade-transition');
export const RSlideXTransitionGroup = createSimpleTransitionGroup('slide-x-transition');
export const RSlideXReverseTransitionGroup = createSimpleTransitionGroup('slide-x-reverse-transition');
export const RSlideYTransitionGroup = createSimpleTransitionGroup('slide-y-transition');
export const RSlideYReverseTransitionGroup = createSimpleTransitionGroup('slide-y-reverse-transition');
export const RScaleTransitionGroup = createSimpleTransitionGroup('scale-transition');

// JavaScript transitions
export const RExpandTransition = createJavaScriptTransition('expand-transition', ExpandTransitionGenerator());

/* istanbul ignore next */
// eslint-disable-next-line max-statements
function install (Vue) {
  // Component transitions
  Vue.component('r-bottom-sheet-transition', RBottomSheetTranstion);
  Vue.component('r-carousel-transition', RCarouselTransition);
  Vue.component('r-carousel-reverse-transition', RCarouselReverseTransition);
  Vue.component('r-dialog-transition', RDialogTransition);
  Vue.component('r-dialog-top-transition', RDialogTopTransition);
  Vue.component('r-dialog-bottom-transition', RDialogBottomTransition);
  Vue.component('r-dialog-left-transition', RDialogLeftTransition);
  Vue.component('r-dialog-right-transition', RDialogRightTransition);
  Vue.component('r-tab-reverse-transition', RTabReverseTransition);
  Vue.component('r-tab-transition', RTabTransition);
  Vue.component('r-fab-transition', RFabTransition);
  Vue.component('r-menu-transition', RMenuTransition);

  // Generic transitions
  Vue.component('r-fade-transition', RFadeTransition);
  Vue.component('r-scale-transition', RScaleTransition);
  Vue.component('r-slide-x-transition', RSlideXTransition);
  Vue.component('r-slide-x-reverse-transition', RSlideXReverseTransition);
  Vue.component('r-slide-y-transition', RSlideYTransition);
  Vue.component('r-slide-y-reverse-transition', RSlideYReverseTransition);

  // Generic group transitions
  Vue.component('r-fade-transition-group', RFadeTransitionGroup);
  Vue.component('r-slide-x-transition-group', RSlideXTransitionGroup);
  Vue.component('r-slide-x-reverse-transition-group', RSlideXReverseTransitionGroup);
  Vue.component('r-slide-y-transition-group', RSlideYTransitionGroup);
  Vue.component('r-slide-y-reverse-transition-group', RSlideYReverseTransitionGroup);
  Vue.component('r-scale-transition-group', RScaleTransitionGroup);

  // JavaScript transitions
  Vue.component('r-expand-transition', RExpandTransition);
}

export default install;
