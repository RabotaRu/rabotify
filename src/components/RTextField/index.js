import RTextField from './RTextField.vue';

/* istanbul ignore next */
RTextField.install = function install (Vue) {
  Vue.component('r-text-field', RTextField);
};

export default RTextField;
