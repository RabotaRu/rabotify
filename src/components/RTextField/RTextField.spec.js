import { test } from '@util/testing';
import Vue from 'vue/dist/vue.common';
import transitions from '@components/transitions';
import RTextField from '@components/RTextField';
import MdAccountBox from '@md-svg-vue/action/MdAccountBox.vue';

Vue.use(transitions);

// eslint-disable-next-line max-statements
test('RTextField.vue', ({ mount }) => {
  it('should render component and match snapshot', () => {
    const wrapper = mount(RTextField);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should pass events to internal input field', () => {
    const keyup = jest.fn();
    const component = {
      render (h) {
        return h(RTextField, { on: { keyUp: keyup }, props: { download: '' }, attrs: {} });
      }
    };
    const wrapper = mount(component);

    const input = wrapper.find('input')[0];
    input.trigger('keyUp', { keyCode: 65 });

    expect(keyup).toBeCalled();
  });

  it('should render aria-label attribute on text field element with label value and no id', () => {
    const wrapper = mount(RTextField, {
      propsData: {
        label: 'Test'
      },
      attrs: {}
    });

    const inputGroup = wrapper.find('input')[0];
    expect(inputGroup.getAttribute('aria-label')).toBe('Test');
  });

  it('should not render aria-label attribute on text field element with no label value or id', () => {
    const wrapper = mount(RTextField, {
      propsData: {
        label: null
      },
      attrs: {}
    });

    const inputGroup = wrapper.find('input')[0];
    expect(inputGroup.element.getAttribute('aria-label')).toBeFalsy();
  });

  it('should not render aria-label attribute on text field element with id', () => {
    const wrapper = mount(RTextField, {
      propsData: {
        label: 'Test'
      },
      attrs: {
        id: 'Test'
      }
    });

    const inputGroup = wrapper.find('input')[0];
    expect(inputGroup.element.getAttribute('aria-label')).toBeFalsy();
  });

  it('should start out as invalid', () => {
    const wrapper = mount(RTextField, {
      propsData: {
        rules: [ (v) => !!v || 'Required' ]
      }
    });

    expect(wrapper.data().valid).toEqual(false);
  });

  it('should start validating on input', async () => {
    const wrapper = mount(RTextField, {});

    expect(wrapper.data().shouldValidate).toEqual(false);
    wrapper.setProps({ value: 'asd' });
    await wrapper.vm.$nextTick();
    expect(wrapper.data().shouldValidate).toEqual(true);
  });

  it('should not start validating on input if validate-on-blur prop is set', async () => {
    const wrapper = mount(RTextField, {
      propsData: {
        validateOnBlur: true
      }
    });

    expect(wrapper.data().shouldValidate).toEqual(false);
    wrapper.setProps({ value: 'asd' });
    await wrapper.vm.$nextTick();
    expect(wrapper.data().shouldValidate).toEqual(false);
  });

  it('should not display counter when set to false', async () => {
    const wrapper = mount(RTextField, {
      propsData: {
        counter: true,
        max: 50
      }
    });

    expect(wrapper.find('.input-group__counter')[0]).not.toBe(undefined);
    expect(wrapper.html()).toMatchSnapshot();

    wrapper.setProps({ counter: false });
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
    expect(wrapper.find('.input-group__counter')[0]).toBe(undefined);
  });

  it('should have readonly attribute', () => {
    const wrapper = mount(RTextField, {
      propsData: {
        readonly: true
      }
    });

    const input = wrapper.find('input')[0];

    expect(input.getAttribute('readonly')).toBe('readonly');
  });

  it('should clear input value', async () => {
    const wrapper = mount(RTextField, {
      propsData: {
        clearable: true,
        value: 'foo'
      }
    });

    const clear = wrapper.find('.input-group__append-icon')[0];
    const input = jest.fn();
    wrapper.vm.$on('input', input);

    expect(wrapper.vm.inputValue).toBe('foo');

    clear.trigger('click');

    await wrapper.vm.$nextTick();

    expect(input).toHaveBeenCalledWith(null);
  });

  it('should not clear input if not clearable and has appended icon (with callback)', async () => {
    const appendIconCb = jest.fn();
    const wrapper = mount(RTextField, {
      propsData: {
        value: 'foo',
        appendIconCb
      },
      slots: {
        appendIcon: [ MdAccountBox ]
      }
    });

    const icon = wrapper.find('.input-group__append-icon')[0];
    icon.trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.inputValue).toBe('foo');
    expect(appendIconCb.mock.calls).toHaveLength(1);
  });

  it('should not clear input if not clearable and has appended icon (without callback)', async () => {
    const wrapper = mount(RTextField, {
      propsData: {
        value: 'foo'
      },
      slots: {
        appendIcon: [ MdAccountBox ]
      }
    });

    const icon = wrapper.find('.input-group__append-icon')[0];
    icon.trigger('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.inputValue).toBe('foo');
  });

  it('should start validating on blur', async () => {
    const wrapper = mount(RTextField, {});

    const input = wrapper.find('input')[0];
    expect(wrapper.data().shouldValidate).toEqual(false);
    input.trigger('focus');
    await wrapper.vm.$nextTick();
    input.trigger('blur');
    await wrapper.vm.$nextTick();
    expect(wrapper.data().shouldValidate).toEqual(true);
  });

  it('should keep its value on blur', async () => {
    const wrapper = mount(RTextField, {
      propsData: {
        value: 'asd'
      }
    });

    const input = wrapper.find('input')[0];

    input.element.value = 'fgh';
    input.trigger('input');
    input.trigger('blur');

    expect(input.element.value).toBe('fgh');
  });

  it('should update if value is changed externally', async () => {
    const wrapper = mount(RTextField, {});

    const input = wrapper.find('input')[0];

    wrapper.setProps({ value: 'fgh' });
    expect(input.element.value).toBe('fgh');

    input.trigger('focus');
    wrapper.setProps({ value: 'jkl' });
    expect(input.element.value).toBe('jkl');
  });

  it('should fire a single change event on blur', async () => {
    let value = 'asd';
    const change = jest.fn();

    const component = {
      render (h) {
        return h(RTextField, {
          on: {
            input: (i) => value = i,
            change
          },
          props: { value }
        });
      }
    };
    const wrapper = mount(component);

    const input = wrapper.find('input')[0];

    input.trigger('focus');
    await wrapper.vm.$nextTick();
    input.element.value = 'fgh';
    input.trigger('input');

    await wrapper.vm.$nextTick();
    input.trigger('blur');
    await wrapper.vm.$nextTick();

    expect(change).toBeCalledWith('fgh');
    expect(change.mock.calls).toHaveLength(1);
  });

  it('should not emit change event if value has not changed', async () => {
    const change = jest.fn();
    let value = 'test';
    const component = {
      render (h) {
        return h(RTextField, {
          on: {
            input: i => value = i,
            change
          },
          props: { value }
        });
      }
    };
    const wrapper = mount(component);

    const input = wrapper.find('input')[0];

    input.trigger('focus');
    await wrapper.vm.$nextTick();
    input.trigger('blur');
    await wrapper.vm.$nextTick();

    expect(change.mock.calls).toHaveLength(0);
  });

  it('should display the number 0', async () => {
    const wrapper = mount(RTextField, {
      propsData: { value: 0 }
    });

    expect(wrapper.vm.$refs.input.value).toBe('0');
  });

  it('should reset internal change on blur', async () => {
    const wrapper = mount(RTextField);

    wrapper.setProps({ value: 'foo' });
    wrapper.vm.internalChange = true;
    await wrapper.vm.$nextTick();

    expect(wrapper.vm.internalChange).toBe(true);
    wrapper.vm.blur();
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.internalChange).toBe(false);
  });

  it('should calculate element height when using auto-grow prop', async () => {
    let value = '';
    const component = {
      render (h) {
        return h(RTextField, {
          on: {
            input: i => value = i
          },
          props: {
            value,
            multiLine: true,
            autoGrow: true
          }
        });
      }
    };

    const wrapper = mount(component);
    const input = wrapper.find('textarea')[0];

    input.trigger('focus');
    await wrapper.vm.$nextTick();
    input.element.value = 'this is a really long text that should hopefully make auto-grow kick in. maybe?';
    input.trigger('input');
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
    expect(input.element.style.getPropertyValue('height').length).not.toBe(0);
  });

  it('should match multi-line snapshot', () => {
    const wrapper = mount(RTextField, {
      propsData: {
        multiLine: true
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match textarea snapshot', () => {
    const wrapper = mount(RTextField, {
      propsData: {
        textarea: true
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should match auto-grow snapshot', async () => {
    const wrapper = mount(RTextField, {
      propsData: {
        textarea: true,
        autoGrow: true
      }
    });
    await wrapper.vm.$nextTick();
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render no-resize the same if already auto-grow', () => {
    const wrappers = [
      { autoGrow: true, multiLine: true },
      { autoGrow: true, textarea: true }
    ].map(propsData => mount(RTextField, { propsData }));

    wrappers.forEach(async wrapper => {
      await wrapper.vm.$nextTick();
      const html1 = wrapper.html();

      wrapper.setProps({ noResize: true });
      // will still pass without this, do not remove
      await wrapper.vm.$nextTick();
      const html2 = wrapper.html();

      expect(html2).toBe(html1);
    });
  });

  it('should always show details when preserveDetails is `true`', async () => {
    const wrapper = mount(RTextField, {
      propsData: {
        value: '123',
        rules: [ (v) => !!v || 'Required' ],
        preserveDetails: true
      }
    });

    expect(wrapper.data().valid).toEqual(true);

    const details = wrapper.find('.input-group__details')[0];
    await wrapper.vm.$nextTick();
    expect(details.element.style.display).not.toBe('none');

    expect(wrapper.html()).toMatchSnapshot();
  });
});
