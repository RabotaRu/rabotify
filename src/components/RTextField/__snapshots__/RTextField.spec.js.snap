// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RTextField.vue should always show details when preserveDetails is \`true\` 1`] = `

<div class="input-group input-group_dirty input-group_text-field">
  <div class="input-group__input">
    <input tabindex="0"
           type="text"
    >
  </div>
  <div class="input-group__details">
  </div>
</div>

`;

exports[`RTextField.vue should calculate element height when using auto-grow prop 1`] = `

<div class="input-group input-group_focused input-group_dirty input-group_text-field input-group_multi-line input-group_no-resize">
  <div class="input-group__input">
    <textarea tabindex="0"
              rows="5"
              style="height: 120px;"
    >
    </textarea>
  </div>
  <div class="input-group__details"
       style="display: none;"
  >
  </div>
</div>

`;

exports[`RTextField.vue should match auto-grow snapshot 1`] = `

<div class="input-group input-group_text-field input-group_textarea input-group_no-resize">
  <div class="input-group__input">
    <textarea tabindex="0"
              rows="5"
              style="height: 124px;"
    >
    </textarea>
  </div>
  <div class="input-group__details"
       style="display: none;"
  >
  </div>
</div>

`;

exports[`RTextField.vue should match multi-line snapshot 1`] = `

<div class="input-group input-group_text-field input-group_multi-line">
  <div class="input-group__input">
    <textarea tabindex="0"
              rows="5"
    >
    </textarea>
  </div>
  <div class="input-group__details"
       style="display: none;"
  >
  </div>
</div>

`;

exports[`RTextField.vue should match textarea snapshot 1`] = `

<div class="input-group input-group_text-field input-group_textarea">
  <div class="input-group__input">
    <textarea tabindex="0"
              rows="5"
    >
    </textarea>
  </div>
  <div class="input-group__details"
       style="display: none;"
  >
  </div>
</div>

`;

exports[`RTextField.vue should not display counter when set to false 1`] = `

<div class="input-group input-group_text-field">
  <div class="input-group__input">
    <input tabindex="0"
           type="text"
    >
  </div>
  <div class="input-group__details">
    <div class="input-group__counter">
      0 / 25
    </div>
  </div>
</div>

`;

exports[`RTextField.vue should not display counter when set to false 2`] = `

<div class="input-group input-group_text-field">
  <div class="input-group__input">
    <input tabindex="0"
           type="text"
    >
  </div>
  <div class="input-group__details"
       style="display: none;"
  >
  </div>
</div>

`;

exports[`RTextField.vue should render component and match snapshot 1`] = `

<div class="input-group input-group_text-field">
  <div class="input-group__input">
    <input tabindex="0"
           type="text"
    >
  </div>
  <div class="input-group__details"
       style="display: none;"
  >
  </div>
</div>

`;
